Docs: Added more details to Operator Precedence section

NHDaly opened the pull request:
From Issue #14933.

- Made Operator Precedence a higher heading, since it's not really a subsection of `##Numeric Comparisons`, and it references all the operators described in earlier sections on the page.
- Added a link to the source file that lists *all* operator precedences.
- Added code examples for `Base.operator_precedence` to show how to programmatically find an operator's precedence.

NHDaly committed:
Added more details to Operator Precedence section

From Issue #14933.

- Made Operator Precedence a higher heading, since it's not really a subsection of `##Numeric Comparisons`, and it references all the operators described in earlier sections on the page.
- Added a link to the source file that lists *all* operator precedences.
- Added code examples for `Base.operator_precedence` to show how to programmatically find an operator's precedence.

stevengj commented on doc/src/manual/mathematical-operations.md:
```diff
@@ -348,16 +359,18 @@ Julia applies the following order of operations, from highest precedence to lowe
-operations are defined over as broad a class of numerical values as permit sensible definitions,
-including integers, floating-point numbers, rationals, and complexes, wherever such definitions
-make sense.
+You can also find the numerical precedence for any given operator via the built-in function `Base.operator_precedence`, where higher numbers take precedence:
```
If we are going to document this function, it should have a docstring too.

NHDaly replied:
Done. See 432e69b.

Sacha0 commented on doc/src/manual/mathematical-operations.md:
```diff
@@ -331,7 +331,18 @@ comparison is undefined. It is strongly recommended not to use expressions with
+
+Julia provides a comprehensive collection of mathematical functions and operators. These mathematical
+operations are defined over as broad a class of numerical values as permit sensible definitions,
+including integers, floating-point numbers, rationals, and complexes, wherever such definitions
```
Perhaps "complex numbers" rather than "complexes"? Or by "complexes" do you mean composite types of some form?

NHDaly replied:
+1 to "complex numbers". 

Should I just make a second PR to clean this language up, or combine it here?

stevengj replied:
Can be fixed in another PR.

NHDaly replied:
#21363

Sacha0 commented on doc/src/manual/mathematical-operations.md:
```diff
@@ -331,7 +331,18 @@ comparison is undefined. It is strongly recommended not to use expressions with
+
+Moreover, these functions (like any Julia function) can be applied in "vectorized" fashion to
+arrays and other collections with the [dot syntax](@ref man-vectorized) `f.(A)`,
+e.g. `sin.(A)` will compute the elementwise sine of each element of an array `A`.
```
Perhaps "compute the sine of each element" rather than "compute the elementwise sine of each element"?

NHDaly replied:
I am completely ambivalent to this language; like you said, I just moved it.

But I think I agree that the wording you suggested seems better. Should I change it, or make a new cleanup PR?

Sacha0 replied:
Followup cleanup PR sounds great as above :).

NHDaly replied:
#21363

Sacha0 commented on doc/src/manual/mathematical-operations.md:
```diff
@@ -348,16 +359,18 @@ Julia applies the following order of operations, from highest precedence to lowe
 
-### Elementary Functions
+For a complete list of *every* Julia operator's precendence, see the top of this file:
+https://github.com/JuliaLang/julia/blob/master/src/julia-parser.scm
```
(Is hyperlink mark up supported? If so, perhaps mark up the reference?)

NHDaly replied:
It *is* supported, but I thought that since this case was a link to code, keeping the url explicit made sense. But I could also change it.

Do you have a suggestion? Maybe just something like this?:

```
[julia/src/julia-parser.scm](https://github.com/JuliaLang/julia/blob/master/src/julia-parser.scm)
```

stevengj replied:
Note the spelling of "precedence"

NHDaly replied:
oops, thanks. Fixed.

Sacha0 replied:
Your suggestion above LGTM :).

NHDaly committed:
Added docstring to `Base.operator_precedence`

As part of #14933, added docstring and jldoctest examples to Base.operator_precedence(s::Symbol).

NHDaly reviewed (commented):
I've added a docstring, and replied to comments. PTAL! :) Thanks!

tkelman commented on base/show.jl:
```diff
@@ -469,6 +469,11 @@ end
+"""
+    operator_precedence(s::Symbol)
+
+Returns an integer representing the precdedence of a Julia operator, relative to other operators. Higher numbers take precedence over lower-numbered operators. Returns `0` if a symbol is not a valid operator.
```
should wrap long lines, including in docstrings

tkelman replied:
precedence is also misspelled

NHDaly replied:
Oops, sorry. That's what I get for trying to use Github's built-in editor. I've fixed it up now. Thanks.

Sacha0 commented on base/show.jl:
```diff
@@ -469,6 +469,14 @@ end
+"""
+    operator_precedence(s::Symbol)
+
+Returns an integer representing the precedence of a Julia operator, relative
```
Perhaps "Return" rather than "Returns", and likewise below? (See the second numbered recommendation a few paragraphs down in https://docs.julialang.org/en/stable/manual/documentation/.)

NHDaly replied:
Done. Thanks, I missed that when reading the documentation documentation last time. 😊

Sacha0 commented on base/show.jl:
```diff
@@ -469,6 +469,14 @@ end
+Returns an integer representing the precedence of a Julia operator, relative
+to other operators. Higher numbers take precedence over lower-numbered
+operators. Returns `0` if a symbol is not a valid operator.
+"""
```
Perhaps "a Julia operator" -> "operator `s`"?

Similarly, perhaps "if a symbol is not a valid operator" -> "if `s` is not a valid operator"?

For consistency, perhaps "Higher-numbered operators take precedence over lower-numbered operators."?

NHDaly replied:
Done. Thanks, that's better.

NHDaly commented on doc/src/manual/mathematical-operations.md:
````diff
@@ -348,16 +359,18 @@ Julia applies the following order of operations, from highest precedence to lowe
+
+julia> Base.operator_precedence(:+=), Base.operator_precedence(:(=))  # (Note the necessary parens on `:(=)`)
+(1,1)
+```
````
Done. I copied them both.

I thought about doing this at first, but I wasn't sure if that's redundant? Lemme know what you think.

NHDaly commented:
Done. PTAL, thanks

Sacha0 commented on base/show.jl:
```diff
@@ -469,6 +469,24 @@ end
+
+Return an integer representing the precedence of operator `s`, relative
+to other operators. Higher numbers take precedence over lower-numbered
+operators. Return `0` if `s` is not a valid operator.
```
Perhaps "Higher-numbered operators take precedence over lower-numbered operators." for consistency? (Fell by the wayside in another comment thread.)

NHDaly replied:
Done. Thanks!

Sacha0 commented on doc/src/manual/mathematical-operations.md:
```diff
@@ -348,16 +359,18 @@ Julia applies the following order of operations, from highest precedence to lowe
 
-### Elementary Functions
+For a complete list of *every* Julia operator's precedence, see the top of this file:
+https://github.com/JuliaLang/julia/blob/master/src/julia-parser.scm
```
Link markup fell by the wayside as well it seems? :)

NHDaly replied:
Oops! Sorry about that. Thankfully, the emails aren't deleted. 😄

Here's what I had last said, and i was waiting for input:

> Do you have a suggestion? Maybe just something like this?:
> ```
>  [julia/src/julia-parser.scm](https://github.com/JuliaLang/julia/blob/master/src/julia-parser.scm)
> ```


I'll do that for now and see what you think! :)

Sacha0 reviewed (approved):
@NHDaly, this looks great! (A couple comments fell through the cracks so I re-commented. If you are not enthusiastic about making those touchups in this PR, don't worry about them: Unless someone else beats me to it, I will hit merge in the next day or so in any case. Much thanks! :) )

NHDaly commented:
> A couple comments fell through the cracks so I re-commented.

Oh, whoops, you're right. It looks like when I force-push an amended commit, it erases the comment threads? I didn't realize that. Sorry! (we don't use github at work, so i still make many github mistakes. :D) Is it preferable, then, to just send up separate commits until everything is all done, and then do any rebasing at the end?

NHDaly commented:
SGTM. I did those changes, so this looks good to submit whenever you're ready. Thanks for your help, @Sacha0! :D

Sacha0 commented:
No worries! You can access the relevant comments (related to outdated commits) by clicking the "Show outdated" links above (on the right). Best!

Sacha0 commented:
Thanks @NHDaly! :)

NHDaly commented:
thanks! :)
